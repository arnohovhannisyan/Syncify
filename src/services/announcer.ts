import { window, commands, env, Uri } from "vscode";
import { localize } from "~/services";

export async function showAnnouncement(url: string): Promise<void> {
	return new Promise(resolve => {
		const message = window.setStatusBarMessage(
			localize("(info) announcementAvailable")
		);

		const openBtn = window.createStatusBarItem(1);
		const dismissBtn = window.createStatusBarItem(1);

		const dispose = () => {
			openDisposable.dispose();
			dismissDisposable.dispose();
			openBtn.dispose();
			dismissBtn.dispose();
			message.dispose();

			resolve();
		};

		const openDisposable = commands.registerCommand(
			"syncify.openAnnouncement",
			() => {
				env.openExternal(Uri.parse(url));

				dispose();
			}
		);

		const dismissDisposable = commands.registerCommand(
			"syncify.dismissAnnouncement",
			dispose
		);

		openBtn.command = "syncify.openAnnouncement";
		openBtn.text = `$(check) ${localize("(label) open")}`;
		openBtn.show();

		dismissBtn.command = "syncify.dismissAnnouncement";
		dismissBtn.text = `$(x) ${localize("(label) dismiss")}`;
		dismissBtn.show();
	});
}
