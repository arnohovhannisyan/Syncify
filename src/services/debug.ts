import { commands } from "vscode";
import actions from "~/redux/actions";
import { store } from "~/redux/store";

export class Debug {
  public static log(...args: any[]) {
    const { isDebugMode } = store.getState();

    if (isDebugMode) console.log(...args);
  }

  public static setDebug(value: boolean) {
    if (value) {
      store.dispatch(actions.enableDebugMode());
    } else {
      store.dispatch(actions.disableDebugMode());
    }

    return commands.executeCommand("syncify.reinitialize");
  }
}
