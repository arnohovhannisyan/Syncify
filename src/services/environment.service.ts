import { normalize, resolve } from "path";
import { state } from "../state";

export class EnvironmentService {
  public locations = {
    userFolder: null
  };

  constructor() {
    const slash = normalize("/");
    if (!process.env.VSCODE_PORTABLE) {
      const path = resolve(state.context.globalStoragePath, "../../..").concat(
        slash
      );
      this.locations.userFolder = resolve(path, "User").concat(slash);
    } else {
      const path = process.env.VSCODE_PORTABLE;
      this.locations.userFolder = resolve(path, "user-data/User").concat(slash);
    }
  }
}
