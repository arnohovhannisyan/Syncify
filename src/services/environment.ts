import { resolve } from "path";
import pkg from "~/../package.json";
import state from "~/state";

export const Environment = {
	pkg,

	get userFolder() {
		const path = process.env.VSCODE_PORTABLE
			? resolve(process.env.VSCODE_PORTABLE, "user-data")
			: resolve(Environment.globalStoragePath, "../../..");

		return resolve(path, "User");
	},

	get repoFolder() {
		return resolve(Environment.globalStoragePath, "repo");
	},

	get settings() {
		return resolve(Environment.globalStoragePath, "settings.json");
	},

	get customFilesFolder() {
		return resolve(Environment.userFolder, "customFiles");
	},

	get vsixFolder() {
		return resolve(Environment.userFolder, "vsix");
	},

	get conflictsFolder() {
		return resolve(Environment.globalStoragePath, "conflicts");
	},

	get globalStoragePath() {
		return state.context?.globalStoragePath ?? "";
	},

	get extensionPath() {
		return state.context?.extensionPath ?? "";
	},

	get os() {
		if (process.platform === "win32") return "windows";
		if (process.platform === "darwin") return "mac";
		return "linux";
	},

	version: pkg.version,

	extensionId: `${pkg.publisher}.${pkg.name}`,

	oauthClientIds: {
		github: "0b56a3589b5582d11832",
		gitlab: "32c563edb04c312c7959fd1c4863e883878ed4af1f39d6d788c9758d4916a0db",
		bitbucket: "zhkr5tYsZsUfN9KkDn"
	}
};
