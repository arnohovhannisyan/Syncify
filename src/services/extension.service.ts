import { commands, extensions } from "vscode";

export class ExtensionService {
  public async installExtension(id: string): Promise<void> {
    return commands.executeCommand("workbench.extensions.installExtension", id);
  }

  public async uninstallExtension(id: string): Promise<void> {
    return commands.executeCommand(
      "workbench.extensions.uninstallExtension",
      id
    );
  }

  public getMissingExtensions(downloadedExtensions: string[]): string[] {
    const installedExtensions = extensions.all
      .filter(ext => !ext.packageJSON.isBuiltin)
      .map(ext => ext.id);
    return downloadedExtensions.filter(
      ext => !installedExtensions.includes(ext)
    );
  }

  public getUnneededExtensions(downloadedExtensions: string[]): string[] {
    const installedExtensions = extensions.all
      .filter(ext => !ext.packageJSON.isBuiltin)
      .map(ext => ext.id);
    return installedExtensions.filter(
      ext => !downloadedExtensions.includes(ext)
    );
  }
}
