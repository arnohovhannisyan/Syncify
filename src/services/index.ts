export * from "~/services/environment.service";
export * from "~/services/localization.service";
export * from "~/services/custom-file.service";
export * from "~/services/extensions.service";
export * from "~/services/factory.service";
export * from "~/services/fs.service";
export * from "~/services/oauth.service";
export * from "~/services/init.service";
export * from "~/services/logger.service";
export * from "~/services/pragma.service";
export * from "~/services/settings.service";
export * from "~/services/utilities.service";
export * from "~/services/watcher.service";
export * from "~/services/webview.service";
