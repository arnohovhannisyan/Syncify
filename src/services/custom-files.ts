import { Environment, FS, localize, Logger } from "~/services";
import { basename, resolve } from "path";
import {
	InputBoxOptions,
	OpenDialogOptions,
	QuickPickItem,
	Uri,
	window,
	workspace,
} from "vscode";

export const importFile = async (uri?: Uri): Promise<void> => {
	try {
		const customFilesFolder = await Environment.customFilesFolder();

		await FS.mkdir(customFilesFolder);

		const allFiles = await FS.listFiles(customFilesFolder, ["**/*"]);

		if (allFiles.length === 0) {
			await window.showInformationMessage(
				localize("(info) CustomFiles.import -> noFiles"),
			);

			return;
		}

		const folder = await (async () => {
			if (uri) return uri.fsPath;

			if (!workspace.workspaceFolders) return;

			if (workspace.workspaceFolders.length === 1) {
				return workspace.workspaceFolders[0].uri.fsPath;
			}

			const result = await window.showQuickPick(
				workspace.workspaceFolders.map<QuickPickItem>((folder) => ({
					label: folder.name,
					description: folder.uri.fsPath,
				})),
				{
					placeHolder: localize(
						"(prompt) CustomFiles.import -> folder placeholder",
					),
				},
			);

			if (!result) return;

			const selectedWorkspace = workspace.workspaceFolders.find(
				(folder) => folder.uri.fsPath === result.description,
			);

			return selectedWorkspace?.uri.fsPath;
		})();

		if (!folder) return;

		const selectedFile = await window.showQuickPick(
			allFiles.map((file) => basename(file)),
			{
				placeHolder: localize(
					"(prompt) CustomFiles.import -> file placeholder",
				),
			},
		);

		if (!selectedFile) return;

		const filepath = resolve(customFilesFolder, selectedFile);

		const options: InputBoxOptions = {
			prompt: localize("(prompt) CustomFiles.import -> file name"),
			value: selectedFile,
		};

		const filename = (await window.showInputBox(options)) ?? selectedFile;

		if (filename.length === 0) return;

		const contents = await FS.readBuffer(filepath);
		await FS.write(resolve(folder, filename), contents);
	} catch (error) {
		void Logger.error(error);
	}
};

const registerDialogOptions: OpenDialogOptions = {
	canSelectMany: false,
	openLabel: localize("(label) CustomFiles.register -> select file"),
};

export const registerFile = async (uri?: Uri): Promise<void> => {
	try {
		const customFilesFolder = await Environment.customFilesFolder();

		await FS.mkdir(customFilesFolder);

		const filepath =
			uri?.fsPath ??
			(await window.showOpenDialog(registerDialogOptions))?.[0]?.fsPath;

		if (!filepath) return;

		const originalFilename = basename(filepath);

		const options: InputBoxOptions = {
			prompt: localize("(prompt) CustomFiles.register -> name"),
			value: originalFilename,
		};

		const filename = (await window.showInputBox(options)) ?? originalFilename;

		if (filename.length === 0) return;

		const newPath = resolve(customFilesFolder, filename);

		if (await FS.exists(newPath)) {
			const result = await window.showWarningMessage(
				localize("(prompt) CustomFiles.register -> file exists"),
				localize("(label) no"),
				localize("(label) yes"),
			);

			if (result !== localize("(label) yes")) return;
		}

		const contents = await FS.readBuffer(filepath);

		await FS.write(newPath, contents);

		await window.showInformationMessage(
			localize("(info) CustomFiles.register -> registered", filename),
		);
	} catch (error) {
		void Logger.error(error);
	}
};
