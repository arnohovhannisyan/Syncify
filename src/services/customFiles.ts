import { basename, resolve } from "path";
import { Uri, window, workspace } from "vscode";
import { Environment, FS, localize } from "~/services";

export class CustomFiles {
  public static async import(uri?: Uri): Promise<void> {
    if (!workspace.workspaceFolders.length) {
      return;
    }

    const folderExists = await FS.exists(Environment.customFilesFolder);

    if (!folderExists) {
      await FS.mkdir(Environment.customFilesFolder);
    }

    const allFiles = await FS.listFiles(Environment.customFilesFolder);

    if (!allFiles.length) {
      await window.showInformationMessage(
        localize("(info) customFiles.noFilesAvailable")
      );
      return;
    }

    const folder = await (async () => {
      if (uri) {
        return uri.fsPath;
      }

      if (workspace.workspaceFolders.length === 1) {
        return workspace.workspaceFolders[0].uri.fsPath;
      }

      const result = await window.showQuickPick(
        workspace.workspaceFolders.map(f => f.name),
        {
          placeHolder: localize("(prompt) customFile.import.folderPlaceholder")
        }
      );

      return workspace.workspaceFolders.filter(f => f.name === result)[0].uri
        .fsPath;
    })();

    if (!folder) {
      return;
    }

    const filename = await window.showQuickPick(
      allFiles.map(f => basename(f)),
      {
        placeHolder: localize("(prompt) customFile.import.filePlaceholder")
      }
    );

    if (!filename) {
      return;
    }

    const filepath = resolve(Environment.customFilesFolder, filename);
    const contents = await FS.readBuffer(filepath);
    await FS.write(resolve(folder, filename), contents);
  }

  public static async register(uri: Uri) {
    const folderExists = await FS.exists(Environment.customFilesFolder);

    if (!folderExists) {
      await FS.mkdir(Environment.customFilesFolder);
    }

    const filepath = uri
      ? uri.fsPath
      : await (async () => {
          const result = await window.showOpenDialog({
            canSelectMany: false,
            openLabel: localize("(action) customFiles.selectFile")
          });

          if (!result) {
            return;
          }

          return result[0].fsPath;
        })();

    if (!filepath) {
      return;
    }

    const filename = basename(filepath);
    const contents = await FS.readBuffer(filepath);
    const newPath = resolve(Environment.customFilesFolder, filename);
    await FS.write(newPath, contents);
    await window.showInformationMessage(
      localize("(info) customFiles.registered", filename)
    );
  }
}
