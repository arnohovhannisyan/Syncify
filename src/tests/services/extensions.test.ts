import { extensions, Uri, commands } from "vscode";
import { Extensions, Environment, FS } from "~/services";
import { getCleanupPath } from "~/tests/getCleanupPath";
import { resolve } from "path";

function setExtensions(exts: string[]): void {
	(extensions.all as any) = exts.map(ext => ({
		id: ext,
		packageJSON: { isBuiltin: false },
		extensionPath: "",
		isActive: false,
		exports: undefined,
		activate: async () => Promise.resolve()
	}));
}

const cleanupPath = getCleanupPath("services/extensions");

const pathToVsix = resolve(cleanupPath, "vsix");

jest.spyOn(Environment, "vsixFolder", "get").mockReturnValue(pathToVsix);

test("missing extensions", () => {
	setExtensions(["publisher1.extension1"]);

	const missing = Extensions.getMissing([
		"publisher1.extension1",
		"publisher2.extension2",
		"publisher3.extension3"
	]);

	expect(missing).toMatchSnapshot();
});

test("unneeded extensions", () => {
	setExtensions([
		"publisher1.extension1",
		"publisher2.extension2",
		"publisher3.extension3"
	]);

	const unneeded = Extensions.getUnneeded(["publisher1.extension1"]);

	expect(unneeded).toMatchSnapshot();
});

describe("install", () => {
	test("marketplace", async () => {
		const spy = jest.spyOn(commands, "executeCommand");

		await Extensions.install("test.extension");

		expect(spy).toHaveBeenCalledWith(
			"workbench.extensions.installExtension",
			"test.extension"
		);

		spy.mockRestore();
	});

	test("vsix", async () => {
		await FS.mkdir(pathToVsix);

		const spy = jest.spyOn(commands, "executeCommand");

		await FS.write(resolve(pathToVsix, "test.extension.vsix"), "test");

		await Extensions.install("test.extension");

		expect(spy).toHaveBeenCalledWith(
			"workbench.extensions.installExtension",
			Uri.file("")
		);

		spy.mockRestore();

		await FS.remove(pathToVsix);
	});
});

test("uninstall", async () => {
	const spy = jest.spyOn(commands, "executeCommand");

	await Extensions.uninstall("test.extension");

	expect(spy).toHaveBeenCalledWith(
		"workbench.extensions.uninstallExtension",
		"test.extension"
	);

	spy.mockRestore();
});
