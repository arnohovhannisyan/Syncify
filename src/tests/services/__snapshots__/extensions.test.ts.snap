// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`missing extensions 1`] = `
Array [
  "publisher2.extension2",
  "publisher3.extension3",
]
`;

exports[`unneeded extensions 1`] = `
Array [
  "publisher2.extension2",
  "publisher3.extension3",
]
`;
